src/main/scala/net/cakesolutions/CakeBuildInfoPlugin.scala

```
@@ -4,30 +4,42 @@ package net.cakesolutions
 
 import scala.util._
 
-import sbt._
-import sbt.IO._
 import sbt.Keys._
+import sbt._
+import sbtbuildinfo.BuildInfoPlugin
+import sbtbuildinfo.BuildInfoPlugin.autoImport._
 
-import sbtbuildinfo.BuildInfoPlugin, BuildInfoPlugin.autoImport._
-
-// kicks in if the Project uses the BuildInfoPlugin
+/**
+  * Assumes that `git` is installed. Used to configure and load the `BuildInfo` object at runtime. This
+  * object contains information regarding the build such as the git commit hash, etc.
+  *
+  * Configuration of this plugin should be avoided in local project SBT build files.
+  */
 object CakeBuildInfoPlugin extends AutoPlugin {
-  override def requires = BuildInfoPlugin
-  override def trigger = allRequirements
+  require(
+    Try("git --version".!!).isSuccess,
```

`require` in an `object` is a very bad idea.

---

if you want to do a check like this, do it when calculating `version` or something like that.

---

True, but git not being present is also a bad idea - hence why we should handle this error scenario. Either way, SBT needs to stop here, so what is the best way of handing this type of error scenario?

We can sys.exit, throw an exception or appeal to some SBT error handling.

I'm not immediately aware how to handle SBT errors in this type of context, hence why I prefer an exception trace over a sys.exit.

---

you don't necessarily get a stacktrace if you do a side effect in an `object`. If you put the exception in a `Task` or `Setting` calculation you'll get the effect you want.

---

Looking at SBT source for `AutoPlugins`, I can't see any obvious mechanism for handling plugin errors here - other than throwing exceptions or sys.exit.

---

Ah, just seen `AutoPluginException` - I guess we should simply throw this exception rather than allow `requires` to throw the `IllegalArgumentException `.

---

I'm not sure if the `scripted` tests can pass if you do this. They don't run in a git environment.